map.go

```diff
@@ -0,0 +1,118 @@
+//
+// NOTE(x): In case of operating on big amounts of data or need of extended functionality - consider to use https://github.com/streamrail/concurrent-map
+type ConcurrentMap struct {
+	items map[interface{}]interface{}
+	lock  *sync.RWMutex
+}
+
+// Private factory. It assigns items and set up RWMutex
+func newConcurrentMap(items map[interface{}]interface{}) *ConcurrentMap {
+	return &ConcurrentMap{items: items, lock: &sync.RWMutex{}}
+}
+
+// Generic factory. Instantiates and initializes ConcurrentMap with `initCap` capacity.
+func New(initCap int) *ConcurrentMap {
+	items := make(map[interface{}]interface{}, initCap)
+	return newConcurrentMap(items)
+}
+
+// Makes Concurrent copy of the `m`.
+func MakeConcurrentCopy(m map[interface{}]interface{}) *ConcurrentMap {
+	items := make(map[interface{}]interface{}, len(m))
+	for key, value := range m {
+		items[key] = value
+	}
+	return newConcurrentMap(items)
+}
+
+// Makes Concurrent copy of the `m` recursively.
+// In case the value is of type map[interface{}]interface{} - it converts it into ConcurrentMap recursively as well.
+func MakeRecursivelyConcurrentCopy(m map[interface{}]interface{}) *ConcurrentMap {
+	items := make(map[interface{}]interface{}, len(m))
+	for key, value := range m {
+		if x, ok := (value).(map[interface{}]interface{}); ok {
+			items[key] = MakeRecursivelyConcurrentCopy(x)
+		} else {
+			items[key] = value
+		}
+	}
+	return newConcurrentMap(items)
+}
+
+// Retrieves an element from map under given key.
+// Returns false in case there is no entry associated with the key.
+func (this *ConcurrentMap) Get(key interface{}) (interface{}, bool) {
+	this.lock.RLock()
+	defer this.lock.RUnlock()
+
+	val, ok := this.items[key]
+	return val, ok
+}
+
+// Sets the given value under the specified key.
+func (this *ConcurrentMap) Set(key interface{}, val interface{}) {
+	this.lock.Lock()
+	defer this.lock.Unlock()
+
+	if this.items == nil {
+		// we would need atleast one element in map
+		this.items = make(map[interface{}]interface{}, 1)
+	}
+
+	this.items[key] = val
+}
+
+// Sets the given value under the specified key and returns true, if the key didn't exist uppon invokation.
+// Returns false and does nothing, in case there is already an entry with the same key.
+func (this *ConcurrentMap) SetIfNotExists(key interface{}, val interface{}) bool {
+	this.lock.Lock()
+	defer this.lock.Unlock()
+
+	if _, ok := this.items[key]; !ok {
+		if this.items == nil {
+			// we would need atleast one element in map
+			this.items = make(map[interface{}]interface{}, 1)
+		}
+		this.items[key] = val
+		return true
+	}
+	return false
+}
+
+// Removes an element from the map.
+func (this *ConcurrentMap) Remove(key interface{}) {
+	this.lock.Lock()
+	defer this.lock.Unlock()
+
+	delete(this.items, key)
+}
+
+// Returns copy of content as non concurrent(general) map.
+func (this *ConcurrentMap) Items() map[interface{}]interface{} {
```

Comment:
Yes. Actually it is intended to return only Items as non concurrent map. Conversion of any item - is a part of users code. In case of high demand, it could be done as separate method.